File: 1월 5주차/사칙연산.js
function solution(arr) {
    var answer = 1;
    let n = parseInt(arr.length/2) + 1
    console.log(n)
    const mndp = Array(n).fill(0)
    const mxdp = Array(n).fill(0)
    mndp.forEach((val,idx)=>{
        mndp[idx] = Array(n).fill(1/0)
        mxdp[idx] = Array(n).fill(-1/0)
    })
    for(let i=0; i<n; i++){
        mndp[i][i] = arr[i*2]*1
        mxdp[i][i] = arr[i*2]*1
    }
    const operator = []
    for (let i=0; i<arr.length; i++){
        if (arr[i] == '-' || arr[i] == '+'){
            operator.push(arr[i])
        }
    }
    
    for(let calc=0; calc<n; calc++){
        for (let x=0; x<n-calc; x++){
            let y = x+calc
            for (let k=0; k<y; k++){
                if (operator[k] == "-"){
                    mndp[x][y] = Math.min(mndp[x][k] - mxdp[k+1][y], mndp[x][y])
                    mxdp[x][y] = Math.max(mxdp[x][k] - mndp[k+1][y], mxdp[x][y])
                } else{
                    mndp[x][y] = Math.min(mndp[x][k] + mndp[k+1][y],mndp[x][y])
                    mxdp[x][y] = Math.max(mxdp[x][k] + mxdp[k+1][y], mxdp[x][y])
                }    
            }
        }
    }
    
    return mxdp[0][n-1];
}

File: 2월 3주차/사칙연산.py
def solution(strs):
    su = [0]*(len(strs)//2+1)
    n = len(su)
    for i in range(0,len(strs),2):
        su[i//2] = int(strs[i])
    operator = []
    for i in range(1,len(strs),2):
        operator.append(strs[i])
    
    mindp = [[1e9]*n for i in range(n)]
    maxdp = [[-1e9]*n for i in range(n)]
    for i in range(n):
        mindp[i][i] = su[i]
        maxdp[i][i] = su[i]
    
    for calc in range(1,n):
        for x in range(n-calc):
            y = x+calc
            for k in range(y):
                if operator[k] == '+':                
                    maxdp[x][y] = max(maxdp[x][k] + maxdp[k+1][y], maxdp[x][y])
                    mindp[x][y] = min(mindp[x][k] + mindp[k+1][y], mindp[x][y])
                else:
                    maxdp[x][y] = max(maxdp[x][k] - mindp[k+1][y], maxdp[x][y])
                    mindp[x][y] = min(mindp[x][k] - maxdp[k+1][y], mindp[x][y])
                    
    return maxdp[0][n-1]